assets/src/edit-story/components/panels/textStyle/useRichTextFormatting.js

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useMemo, useCallback } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import generatePatternStyles from '../../../utils/generatePatternStyles';
+import useRichText from '../../richText/useRichText';
+import {
+  getHTMLInfo,
+  toggleBoldInHTML,
+  setFontWeightInHTML,
+  toggleItalicInHTML,
+  toggleUnderlineInHTML,
+  setLetterSpacingInHTML,
+  setColorInHTML,
+} from '../../richText/htmlManipulation';
+import { MULTIPLE_VALUE } from '../../form';
+
+function isEqual(a, b) {
+  const isPattern = typeof a === 'object' && (a.type || a.color);
+  if (!isPattern) {
+    return a === b;
+  }
+
+  const aStyle = generatePatternStyles(a);
+  const bStyle = generatePatternStyles(b);
+  const keys = Object.keys(aStyle);
+  return keys.every((key) => aStyle[key] === bStyle[key]);
+}
+
+function reduceWithMultiple(reduced, info) {
+  return Object.fromEntries(
+    Object.keys(info).map((key) => {
+      const wasMultiple = reduced[key] === MULTIPLE_VALUE;
+      const hadValue = typeof reduced[key] !== 'undefined';
+      const areDifferent = hadValue && !isEqual(reduced[key], info[key]);
+      if (wasMultiple || areDifferent) {
+        return [key, MULTIPLE_VALUE];
+      }
+      return [key, info[key]];
+    })
+  );
+}
+
+function useRichTextFormatting(selectedElements, pushUpdate) {
+  const {
+    state: { hasCurrentEditor, selectionInfo },
+    actions: {
+      selectionActions: {
+        toggleBoldInSelection,
+        setFontWeightInSelection,
+        toggleItalicInSelection,
+        toggleUnderlineInSelection,
+        setLetterSpacingInSelection,
+        setColorInSelection,
+      },
+    },
+  } = useRichText();
+
+  const textInfo = useMemo(() => {
+    if (hasCurrentEditor) {
+      return selectionInfo;
+    }
+
+    // loop over all elements, find info for content and reduce to common value
+    // (setting MULTIPLE_VALUE appropriately)
+    return selectedElements
+      .map(({ content }) => content)
+      .map(getHTMLInfo)
+      .reduce(reduceWithMultiple, {});
+  }, [hasCurrentEditor, selectionInfo, selectedElements]);
+
+  const push = useCallback(
+    (updater, ...args) =>
+      pushUpdate(
+        ({ content }) => ({ content: updater(content, ...args) }),
+        true
```

Comment:
I now updates the text field as soon as a value is entered. But the new content is (as always) not submitted until the edit mode is exited.

Reply:
Can you explain how this will go for letter-spacing? The sequence of input:

1. `150` (initial value, indicating `150%`)
2. `15` (the right-most character is deleted)
3. `1` (the right-most character is deleted)
4. `''` (the right-most character is deleted, this is not an empty string)

Reply:
The input handling in the input component here is not optimal - it'll push its changes on every change, and then re-read the current value from the real input, resulting in empty value not being allowed, so it's un-intuitive to type some values (e.g. you cannot clear the text field). That's an optimization of the input component though, not the text formatter IMHO.

Reply:
Ok. Let's proceed and we'll need to see how we can follow up on this. For now this is one of few items that allows intermediate values. It might be best to do this update via the form's `usePresubmitHandler`, but can be done as a follow up.